### DISCUSSION TITLE
AWS CodeCommit: unable to clone , error: reference delta not found

### DISCUSSION THREAD

**Discussion by harshadhulipala5** [fluxcd/flux2#948] [Q&A] [answered]
flux create source git samplerepo --url=https://git-codecommit.ap-southeast-2.amazonaws
.com/v1/repos/helmrelease- --branch=master --namespace=operations --interval=1m --secret-ref=code-commit-secret
✔ authentication configured
✚ generating GitRepository source
► applying GitRepository source
✔ GitRepository source created
◎ waiting for GitRepository source reconciliation
✗ unable to clone 'https://git-codecommit.ap-southeast-2.amazonaws.com/v1/repos/maas-helmrelease', error: reference delta not found

**Comment by harshadhulipala5** [comment]
not sure why i'm unable to clone the repo

**Reply by stefanprodan** [reply]
Can you please try with —git-implementation=libgit2 and see if anything changes

**Reply by harshadhulipala5** [reply]
flux create source git sample-helmrelease \
--url=https://git-codecommit.ap-southeast-2.amazonaws.com/v1/repos/sample-helmrelease \
--branch=master \
--namespace=operations \
--interval=1m \
--secret-ref=code-commit-secret \
--git-implementation=libgit2
✔ authentication configured
✚ generating GitRepository source
► applying GitRepository source
✔ GitRepository source created
◎ waiting for GitRepository source reconciliation
✗ unable to clone 'https://git-codecommit.ap-southeast-2.amazonaws.com/v1/repos/sample-helmrelease', error: reference delta not found

didn't work @stefanprodan

**Reply by harshadhulipala5** [reply]
i was able to clone the existing repo normally but with gitrepository i'm having issue unable to clone

**Reply by harshadhulipala5** [reply]
Hi @stefanprodan 

I was able to create the git repo on flux 

gitrepository is not picking the repo(https://git-codecommit.ap-southeast-2.amazonaws.com/v1/repos/sample-helmrelease) which is have more that one directory
I've delete all and kept only one and created the gitrepo and voila it worked 
later i've added the directories to actual repo and it got sync to git repo

**Reply by stefanprodan** [reply]
Hmm `reference delta not found` means that Flux wasn't able to resolve the HEAD commit. Strange that now it works, I assume git history is different now and that's why it can checkout the latest commit.

**Reply by rossigee** [reply]
The `reference delta not found` error appears to be originating from the `go-git` library. So, I tried switching to:

```
gitImplementation: libgit2
```

And now I no longer get the error in the logs. Things *seem* to be working now...

```
  conditions:
  - lastTransitionTime: "2021-03-14T02:16:42Z"
    message: 'Fetched revision: master/b5b27a92fd8faa16aa724402a4fea2ba13029ad2'
    reason: GitOperationSucceed
    status: "True"
    type: Ready
```

Maybe worth adding to the docs?

**Comment by FeargusOG** [comment]
I'm seeing something similar, except it's a state that flux gets into while trying to reconcile:

```
{"level":"error","ts":"2021-03-10T16:07:02.992Z","logger":"controller.gitrepository","msg":"Reconciler error","reconciler group":"source.toolkit.fluxcd.io","reconciler kind":"GitRepository","name":"flux-system","namespace":"flux-system","error":"unable to clone 'https://git-codecommit.us-east-2.amazonaws.com/v1/repos/<repo-name>', error: reference delta not found"}
```

Everything will be working fine and then I'll push a commit and this will happen. If I revert the commit it goes back to working. I can then push new commits and it works fine. I can even break up the problematic commit into smaller commits and it'll work. So it really doesn't seem to be an issue with the repo.

Is it possible to get more details from the GitRepository controller about the issue?

**Reply by FeargusOG** [reply]
For further reference, here is the (abridged) manifest for our GitRepository:

```
apiVersion: source.toolkit.fluxcd.io/v1beta1
kind: GitRepository
metadata:
  annotations:
    reconcile.fluxcd.io/requestedAt: "2021-03-10T14:35:03.120563Z"
  creationTimestamp: "2021-03-08T14:33:41Z"
  finalizers:
  - finalizers.fluxcd.io
  generation: 1
  labels:
    kustomize.toolkit.fluxcd.io/checksum: 702b5b6a6a31e3b95ee9dd4093baad736bf71202
    kustomize.toolkit.fluxcd.io/name: flux-system
    kustomize.toolkit.fluxcd.io/namespace: flux-system
  name: flux-system
  namespace: flux-system
  resourceVersion: "41664451"
  selfLink: /apis/source.toolkit.fluxcd.io/v1beta1/namespaces/flux-system/gitrepositories/flux-system
  uid: 7a63d79c-9ba1-4ccd-a375-9f984ba663f8
spec:
  gitImplementation: go-git
  interval: 1m0s
  ref:
    branch: master
  secretRef:
    name: flux-system
  timeout: 20s
  url: https://git-codecommit.us-east-2.amazonaws.com/v1/repos/<repo-name>
status:
  artifact:
    checksum: 110d26f99accb9b36d9a43a6d6fc56402fe92499
    lastUpdateTime: "2021-03-10T14:31:31Z"
    path: gitrepository/flux-system/flux-system/4ab30344d5ddc428ca97de7fa965c5e3d66cdba1.tar.gz
    revision: master/4ab30344d5ddc428ca97de7fa965c5e3d66cdba1
    url: http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/4ab30344d5ddc428ca97de7fa965c5e3d66cdba1.tar.gz
  conditions:
  - lastTransitionTime: "2021-03-10T14:34:39Z"
    message: 'unable to clone ''https://git-codecommit.us-east-2.amazonaws.com/v1/repos/<repo-name>'',
      error: reference delta not found'
    reason: GitOperationFailed
    status: "False"
    type: Ready
  lastHandledReconcileAt: "2021-03-10T14:35:03.120563Z"
  observedGeneration: 1
  url: http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/latest.tar.gz
```

**Reply by stefanprodan** [reply]
Use `gitImplementation: libgit2`

**Comment by bburky** [comment]
I think this is the upstream go-git issue: https://github.com/go-git/go-git/issues/336. There appears to be a draft PR to resolve it: https://github.com/go-git/go-git/pull/111

**Reply by bburky** [reply]
https://github.com/go-git/go-git/pull/111 was superseded by https://github.com/go-git/go-git/pull/392, which has been merged into go-git. However there have not been any new releases of go-git since this PR was merged.

**Comment by fbonin** [comment]
—git-implementation=libgit2 definitely fixed the issue for me, however it is not possible to use this flag with the bootstrap command. 

Thus, I was not able to bootstrap an existing repo with a lot of files. I had to bootstrap an empty repo and then add back my files after modifying the file gotk-patches.yaml

**Comment by stefanprodan** [comment, marked as answer]
We have published on the documentation website a guide on how to manually bootstrap a cluster with AWS CodeCommit. 
https://fluxcd.io/docs/use-cases/aws-codecommit/
